File: 백준/Silver/15657. N과 M （8）/N과 M （8）.py
n,m = map(int, input().split())
arr = list(map(int,input().split()))
arr.sort()

ans = [0] * m
v = [0] * n

def dfs(x, start):
    if x == m:
        print(*ans)
        return
    for i in range(start, n):
        ans[x] = arr[i]
        dfs(x+1, i)
        # ans[x] = 0


dfs(0,0)

File: 백준/Silver/15657. N과 M （8）/N과 M （8）.java

import java.util.Arrays;
import java.util.Scanner;

public class Main {
  static int n, m;
  static int arr[];
  static int ans[];
  static boolean visited[];
  static StringBuilder sb = new StringBuilder();

  public static void main(String[] args) {
    Scanner sc = new Scanner(System.in);
    n = sc.nextInt();
    m = sc.nextInt();
    arr = new int[n];
    ans = new int[m];

    for (int i = 0; i < n; i++) {
      arr[i] = sc.nextInt();
    }
    Arrays.sort(arr);

    recur(0, 0);
    System.out.println(sb);
  }
  static void recur(int dept, int start) {
    if (dept == m){
      for (int i = 0; i < m; i++) {
        sb.append(ans[i] + " ");
      }
      sb.append("\n");
      return;
    }
    for (int i = start; i < n; i++){
      ans[dept] = arr[i];
      recur(dept + 1, i);
    }
  }
}
